tests/components/wemo/test_wemo_device.py

```diff
@@ -148,3 +153,64 @@ async def test_async_update_data_subscribed(
     pywemo_device.get_state.reset_mock()
     await device._async_update_data()
     pywemo_device.get_state.assert_not_called()
+
+
+class TestInsight:
+    """Tests specific to the WeMo Insight device."""
+
+    @pytest.fixture
+    def pywemo_model(self):
+        """Pywemo Dimmer models use the light platform (WemoDimmer class)."""
+        return "Insight"
+
+    @pytest.fixture(name="pywemo_device")
+    def pywemo_device_fixture(self, pywemo_device):
+        """Fixture for WeMoDevice instances."""
+        pywemo_device.insight_params = {
+            "currentpower": 1.0,
+            "todaymw": 200000000.0,
+            "state": 0,
+            "onfor": 0,
+            "ontoday": 0,
+            "ontotal": 0,
+            "powerthreshold": 0,
+        }
+        yield pywemo_device
+
+    @pytest.mark.parametrize(
+        "subscribed,state,expected",
+        [
+            (False, 0, True),
+            (False, 1, True),
+            (True, 0, True),
+            (True, 1, False),
+        ],
+    )
+    async def test_should_poll(
+        self,
+        hass,
+        subscribed,
+        state,
+        expected,
+        wemo_entity,
+        pywemo_device,
+        pywemo_registry,
+    ):
+        """Validate the should_poll returns the correct value."""
+        pywemo_registry.is_subscribed.return_value = subscribed
+        pywemo_device.reset_mock()
+        pywemo_device.get_state.return_value = state
+        await async_setup_component(hass, HA_DOMAIN, {})
+        await hass.services.async_call(
```

Comment:
I'd test this by moving time forward. If the case should poll there should be a call to the mock device. If the case should not poll, there should be no call to the mock device.

https://github.com/home-assistant/core/blob/6a93f5b7ad1fc751be8ac3a57aea10f9858a7b6e/tests/common.py#L372-L376

Reply:
Thanks. I switched to using `async_fire_time_changed`. I believe I still need to check for the call to the mock device in the same way, looking for the `force_update=True` argument to `get_state`. That's what indicates the polling took place.

https://github.com/home-assistant/core/blob/18c03e2f8db172d7f6386a59f3364705b6ebb87b/homeassistant/components/wemo/wemo_device.py#L95-L103

Reply:
What is making the other calls to the same method?

Reply:
Updated the test to pass in the expected calls.

Each time the `_async_update_data` method is called, there are at least three calls to get_state(); one inside should_poll, and one inside the Switch and Binary Sensor that is listening for updates to from the coordinator (I'm not sure why they don't happen inside the two Sensors too). These calls happen after the mock is reset and time is moved forward.

```
  File "homeassistant/helpers/update_coordinator.py", line 134, in _handle_refresh_interval
    await self._async_refresh(log_failures=True, scheduled=True)
  File "homeassistant/helpers/update_coordinator.py", line 187, in _async_refresh
    self.data = await self._async_update_data()
  File "homeassistant/components/wemo/wemo_device.py", line 103, in _async_update_data
    if not self.should_poll:
  File "homeassistant/components/wemo/wemo_device.py", line 91, in should_poll
    if isinstance(self.wemo, Insight) and self.wemo.get_state() == 0:
```

```
  File "homeassistant/helpers/update_coordinator.py", line 134, in _handle_refresh_interval
    await self._async_refresh(log_failures=True, scheduled=True)
  File "homeassistant/helpers/update_coordinator.py", line 265, in _async_refresh
    update_callback()
  File "homeassistant/components/wemo/entity.py", line 77, in _handle_coordinator_update
    super()._handle_coordinator_update()
  File "homeassistant/helpers/update_coordinator.py", line 325, in _handle_coordinator_update
    self.async_write_ha_state()
  File "homeassistant/helpers/entity.py", line 464, in async_write_ha_state
    self._async_write_ha_state()
  File "homeassistant/helpers/entity.py", line 498, in _async_write_ha_state
    state = self._stringify_state()
  File "homeassistant/helpers/entity.py", line 470, in _stringify_state
    state = self.state
  File "homeassistant/components/binary_sensor/__init__.py", line 178, in state
    return STATE_ON if self.is_on else STATE_OFF
  File "homeassistant/components/wemo/binary_sensor.py", line 66, in is_on
    return super().is_on and self.wemo.insight_params["state"] == "1"
  File "homeassistant/components/wemo/binary_sensor.py", line 44, in is_on
    return self.wemo.get_state()
```

```
  File "homeassistant/helpers/update_coordinator.py", line 134, in _handle_refresh_interval
    await self._async_refresh(log_failures=True, scheduled=True)
  File "homeassistant/helpers/update_coordinator.py", line 265, in _async_refresh
    update_callback()
  File "homeassistant/components/wemo/entity.py", line 77, in _handle_coordinator_update
    super()._handle_coordinator_update()
  File "homeassistant/helpers/update_coordinator.py", line 325, in _handle_coordinator_update
    self.async_write_ha_state()
  File "homeassistant/helpers/entity.py", line 464, in async_write_ha_state
    self._async_write_ha_state()
  File "homeassistant/helpers/entity.py", line 498, in _async_write_ha_state
    state = self._stringify_state()
  File "homeassistant/helpers/entity.py", line 470, in _stringify_state
    state = self.state
  File "homeassistant/helpers/entity.py", line 882, in state
    return STATE_ON if self.is_on else STATE_OFF
  File "homeassistant/components/wemo/switch.py", line 144, in is_on
    return self.wemo.get_state()
```